File: Semester 4/Class14_Computer_Contest_Level_3_Notes_and_Homework_Dec_10/lvl3_hw_sol/absolutely_acidic.java
import java.util.*;
import java.io.*;
public class homework {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st; 
    public static void main(String[] args) throws IOException{
    	int n = readInt(), freq[] = new int[1001];
    	int mx = 0, mx2 = 0;
    	for(int i=1; i<=n; i++) {
    		int x = readInt(); freq[x] ++;
    		mx = Math.max(mx, freq[x]);
    	}
    	List<Integer> t = new ArrayList();
    	for(int i=1; i<freq.length; i++) {
    		if(freq[i] == mx) t.add(i);
    		if(freq[i] < mx && freq[i] > mx2) mx2 = freq[i];
    	}
    	if(t.size() != 1) {
    		System.out.println(t.get(t.size()-1) - t.get(0));
    		return;
    	}
    	int high = t.get(0);  t.clear();
    	for(int i=1; i<freq.length; i++) {
    		if(freq[i] == mx2) t.add(i);
    	}
    	int small = Math.abs(high - t.get(0));
    	int big = Math.abs(high - t.get(t.size()-1));
    	System.out.println(Math.max(small, big));
    	
    }
    static String next () throws IOException {
        while (st == null || ! st.hasMoreTokens())
            st = new StringTokenizer(br.readLine().trim());
        return st.nextToken();
    }
    static long readLong () throws IOException {
        return Long.parseLong(next());
    }
    static int readInt () throws IOException {
        return Integer.parseInt(next());
    }
    static double readDouble () throws IOException {
        return Double.parseDouble(next());
    }
    static char readCharacter () throws IOException {
        return next().charAt(0);
    }  
    static String readLine () throws IOException {
        return br.readLine().trim();
    }
}

File: Semester 4/Class14_Computer_Contest_Level_3_Notes_and_Homework_Dec_10/lvl3_hw_sol/absolutely_acidic.py
import sys
input = sys.stdin.readline

maxv = 1000
freq = [0]*(maxv + 1)
for i in range(int(input())):
    freq[int(input())] += 1
mx1 = max(freq)
t = []
for i in range(len(freq)):
    if freq[i] == mx1:
        t.append(i)
if len(t) >= 2:
    print(max(t) - min(t))
else:
    high = t[0]
    mx2 = 0
    t2 = []
    for i in range(len(freq)):
        if freq[i] < mx1 and freq[i] > mx2:
            mx2 = freq[i]
    for i in range(len(freq)):
        if freq[i] == mx2:
            t2.append(i)
    t2.sort()
    print(max(abs(high - t2[0]), abs(high - t2[-1])))